runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,158 @@
+local M = {}
+local util = require "vim.lsp.util"
+local ns = vim.api.nvim_create_namespace("lsp-semantic-tokens")
+
+local semantic_tokens = {}
+
+local function modifiers_from_number(x, modifiers_table)
+  local function get_bit(n, k)
+    -- Based on/from https://stackoverflow.com/a/26230537
+    -- (n >> k) & 1
+    return _G.bit.band(_G.bit.rshift(n, k), 1)
+  end
+
+  local modifiers = {}
+  for i = 0, #modifiers_table - 1 do
+    local bit = get_bit(x, i)
+    if bit == 1 then
+      table.insert(modifiers, 1, modifiers_table[i + 1])
+    end
+  end
+
+  return modifiers
+end
+
+M.token_map = {
+  type = "LspType",
+  class = "LspClass",
+  enum = "LspEnum",
+  interface = "LspInterface",
+  struct = "LspStruct",
+  typeParameter = "LspTypeParameter",
+  parameter = "LspParameter",
+  variable = "LspVariable",
+  property = "LspProperty",
+  enumMember = "LspEnumMember",
+  event = "LspEvent",
+  ["function"] = "LspFunction",
+  method = "LspMethod",
+  macro = "LspMacro",
+  keyword = "LspKeyword",
+  modifier = "LspModifier",
+  comment = "LspComment",
+  string = "LspString",
+  number = "LspNumber",
+  regexp = "LspRegexp",
+  operator = "LspOperator"
+}
+
+M.modifiers_map = {
+  declaration = "LspDeclaration",
+  definition = "LspDefinition",
+  readonly = "LspReadonly",
+  static = "LspStatic",
+  deprecated = "LspDeprecated",
+  abstract = "LspAbstract",
+  async = "LspAsync",
+  modification = "LspModification",
+  documentation = "LspDocumentation",
+  defaultLibrary = "LspDefaultLibrary"
+}
+
+local function highlight(buf, token, hl)
+  vim.highlight.range(buf, ns, hl, {token.line, token.start_char}, {token.line, token.start_char + token.length})
+  vim.schedule_wrap(function()
+    vim.api.nvim_buf_set_extmark(
+      buf,
+      ns,
+      token.line,
+      token.start_char,
+      {
+        end_line = token.line,
+        end_col = token.start_char + token.length,
+        hl_group = hl,
+        ephemeral = false,
```

Comment:
TODO: what is the best strategy for highlighting? Probably, we'd assume to do this request not to often. tree-sitter and regex highlight are for highlighting on every keystroke. LSP add's extra semantic fancyness when document is stale, LSP would even notify us with `refresh` when it would be a good moment to refresh.